### PR TITLE
Copter: simplify gcs failsafe, add control_mode_reason

### PR DESCRIPTION
**Pull request by jschall**
Intend to use this state to enhance logging and to help implement smarter failsafe logic such as GPS glitch.

### CODE REVIEW

**Comment by jschall** [comment]
logging added

File: ArduCopter/control_flip.cpp [hunk 1]
```diff
@@ -199,9 +201,9 @@ void Copter::flip_run()
         // check for successful recovery
         if (fabsf(recovery_angle) <= FLIP_RECOVERY_ANGLE) {
             // restore original flight mode
-            if (!set_mode(flip_orig_control_mode)) {
+            if (!set_mode(flip_orig_control_mode, flip_orig_control_mode_reason)) {
```
Comment (OXINARF) [inline ArduCopter/control_flip.cpp]: I understand the reason for this but I don't agree with it. Imagine the previous mode was AltHold and was changed from GCS, then you enter Flip mode from your Tx, you would then be back to AltHold and log would show it was changed from GCS when that's not true.
Even worse is that if you can't enter the previous mode you enter Stabilize but in that case you say that it was reason unknown. For me, these are the exact same case, the reason for change is the simple flight control.

File: ArduCopter/defines.h [hunk 2]
```diff
@@ -106,6 +106,20 @@ enum autopilot_modes {
     BRAKE =        17   // full-brake using inertial/GPS system, no pilot input
 };
 
+enum mode_reason_t {
+    MODE_REASON_UNKNOWN=0,
```
Comment (OXINARF) [inline ArduCopter/defines.h]: Obviously not my call but I'll leave my opinion. I don't like this reason unknown:

- globally it helps developers being lazy by using unknown for any reason that isn't defined already
- its use now seems to be limited to changes from flight control, I would remove this unknown and add FLIGHT_CONTROL or something like that

You are using 0 for log in Plane/Rover and I would leave it like that, just don't document it so people don't use it

File: ArduCopter/events.cpp [hunk 3]
```diff
@@ -112,52 +116,16 @@ void Copter::failsafe_gcs_check()
...
-                set_mode_RTL_or_land_with_pause();
-            }else{
-                // We have no GPS or are very close to home so we will land
-                set_mode_land_with_pause();
+    if (should_disarm_on_failsafe()) {
+        init_disarm_motors();
+    } else {
+        if (control_mode == AUTO && g.failsafe_gcs == FS_GCS_ENABLED_CONTINUE_MISSION) {
+            // continue mission
+        } else if (g.failsafe_gcs != FS_GCS_DISABLED) {
```
Comment (OXINARF) [inline ArduCopter/events.cpp]: Wouldn't be better to check against FS_GCS_ENABLED_ALWAYS_RTL? If it disabled you never get here, so this if is unnecessary and also checking against FS_GCS_ENABLED_ALWAYS_RTL would make other future options easier to add

File: ArduCopter/flight_mode.cpp [hunk 4]
```diff
@@ -103,9 +104,10 @@ bool Copter::set_mode(uint8_t mode)
     // update flight mode
     if (success) {
         // perform any cleanup required by previous flight mode
-        exit_mode(control_mode, mode);
-        control_mode = mode;
-        DataFlash.Log_Write_Mode(control_mode);
+        exit_mode(control_mode, (control_mode_t)mode);
+        control_mode = (control_mode_t)mode;
```
Comment (OXINARF) [inline ArduCopter/flight_mode.cpp]: These casts are unnecessary right?

**Comment by OXINARF** [comment]
First let me say that I think this is very valuable information to have in a log so I'm glad to see it.

I've gone through it and made some comments inline. Also there's an event change (AP_Notify::events.failsafe_mode_change) that disappears, is this intended?

**Comment by jschall** [comment]
The events change is definitely not intended, thank you. I think set_mode can probably set that flag based on the mode change reason.

MODE_REASON_UNKNOWN is a convenience thing. It's also the reason for the initial mode. Perhaps it can be changed to MODE_REASON_INITIAL_MODE. Anyway, I think the fact that the programmer is forced to provide a mode change reason is enough to prevent laziness.

If we add a MODE_REASON_FLIP_EXIT then it makes certain logic harder. Flip is a bizarro mode.

Then again, I did add a BRAKE_TIMEOUT mode reason, so maybe I should add a FLIP_EXIT.

**Comment by OXINARF** [comment]
Regarding REASON_UNKNOWN and the others: I suggested something like FLIGHT_CONTROL precisely because I was thinking of making it more general when it was this kind of change instead of adding several different reasons, but obviously the more specific the more helpful it can be.
I think that this is only useful if at all times you know the reason for a change in mode. When you have an unknown it can be for anything and then this would be almost useless. As I said this is just my opinion.

**Comment by jschall** [comment]
adding need-rework until issues are addressed and this is rebased

**Comment by rmackay9** [comment]
Looks good to me with a couple of small things.
I'm not sure passing the reason code into the land_init() makes things much better.  It's not a big deal but in some small way I think it moves some of the failsafe logic into the land flight mode when it would be better to leave it where it is.
There's an AP_AHRS_NavEKF change that shouldn't be there maybe?  seems unrelated and not used although I think we should be using this gps glitch detection.

I've put a modifed branch here: https://github.com/rmackay9/rmackay9-ardupilot/tree/jon-reason1

**Comment by OXINARF** [comment]
@rmackay9 I'm not sure if you missed my previous comments or not but I've made some in your version. Some comments I made before were just my opinions so I haven't repeated them.

**Comment by rmackay9** [comment]
@OXINARF ah, great, thank!

**Comment by rmackay9** [comment]
I've resolved all the issues reported by @OXINARF (I think) and the modified version of this PR is sitting in my repo's jon-reason2 branch.  https://github.com/rmackay9/rmackay9-ardupilot/commits/jon-reason2

I'll give it another day or so and if there are no further comments I'll push to master.

**Comment by OXINARF** [comment]
@rmackay9 I think you missed this one: https://github.com/rmackay9/rmackay9-ardupilot/commit/283f937e2c3df5668775f6c441266b7d758f9c53#commitcomment-16995159

I also want to say that I maintain my previous opinions (with special emphasis on https://github.com/ArduPilot/ardupilot/pull/3551#discussion_r50874455), Jon has a different opinion and obviously it's up to you on what's best.

**Comment by rmackay9** [comment]
Ok, merged to master.
@OXINARF, thanks for finding that issue I missed.  I think you're right about the mode-change reason issue for Flip.  It doesn't need to maintain the previous reason, we can just use a new reason to say it's the flight controller itself that's changed it.  I pushed the commits to master before thinking about this enough so I'll make an additional change later.

**Pull request closed without merging**